src/umpire/util/AllocationMap.cpp

```diff
@@ -119,122 +55,115 @@ void RecordList::push_back(AllocationRecord rec)
-                                                 bool end)
-  : m_list(list), m_curr(end ? nullptr : m_list->m_tail)
+RecordList::ConstIterator::ConstIterator()
+  : m_list(nullptr), m_curr(nullptr)
 {
 }
 
-const AllocationRecord& RecordListConstIterator::operator*() const
+RecordList::ConstIterator::ConstIterator(const RecordList* list, iterator_begin)
+  : m_list(list), m_curr(m_list->m_tail)
 {
-  if (!m_curr) UMPIRE_ERROR("Cannot dereference nullptr");
-  return m_curr->rec;
 }
 
-const AllocationRecord* RecordListConstIterator::operator->() const
+RecordList::ConstIterator::ConstIterator(const RecordList* list, iterator_end)
+  : m_list(list), m_curr(nullptr)
 {
-  return m_curr ? &(m_curr->rec) : nullptr;
 }
 
-RecordListConstIterator& RecordListConstIterator::operator++()
+const AllocationRecord&
+RecordList::ConstIterator::operator*()
 {
+  return *operator->();
+}
+
+const AllocationRecord*
+RecordList::ConstIterator::operator->()
+{
+  if (!m_curr) UMPIRE_ERROR("Cannot dereference nullptr");
+  return &m_curr->rec;
+}
+
+RecordList::ConstIterator& RecordList::ConstIterator::operator++()
+{
+  if (!m_curr) UMPIRE_ERROR("Cannot dereference nullptr");
   m_curr = m_curr->prev;
   return *this;
 }
 
-RecordListConstIterator RecordListConstIterator::operator++(int)
+RecordList::ConstIterator RecordList::ConstIterator::operator++(int)
 {
-  RecordListConstIterator tmp{*this};
-  m_curr = m_curr->prev;
+  ConstIterator tmp{*this};
+  this->operator++();
   return tmp;
 }
 
-bool RecordListConstIterator::operator==(const RecordListConstIterator& other)
+bool RecordList::ConstIterator::operator==(const RecordList::ConstIterator& other) const
 {
   return m_list == other.m_list && m_curr == other.m_curr;
 }
 
-bool RecordListConstIterator::operator!=(const RecordListConstIterator& other)
+bool RecordList::ConstIterator::operator!=(const RecordList::ConstIterator& other) const
 {
   return !(*this == other);
 }
 
 // AllocationMap
 AllocationMap::AllocationMap() :
-  m_array(nullptr),
-  m_size(0),
-  m_last(nullptr),
-  m_mutex()
+  m_map{}, m_size{0}, m_mutex{}
 {
-  // Create new judy array
-  m_array = judy_open(judy_max_levels, judy_depth);
-}
-
-AllocationMap::~AllocationMap()
-{
-  // Delete all entries
-  clear();
-
-  // Close the judy array, freeing all memory.
-  judy_close(m_array);
 }
 
 void AllocationMap::insert(void* ptr, AllocationRecord record)
 {
   UMPIRE_LOCK;
   UMPIRE_LOG(Debug, "Inserting " << ptr);
 
-  // Find the key
-  m_last = judy_cell(m_array, reinterpret_cast<unsigned char*>(&ptr), judy_max);
-  UMPIRE_ASSERT(m_last);
+  Map::Iterator iter{m_map.end()};
+  bool found;
 
-  auto plist = reinterpret_cast<RecordList**>(m_last);
-
-  if (!*plist) {
-    // if there is no list there, create one and emplace the record
-    (*plist) = new (list_pool.allocate()) RecordList{record};
-  }
-  else {
-    // else, push onto that list
-    (*plist)->push_back(record);
+  std::tie(iter, found) = m_map.get(ptr, record);
```

Comment:
This crazy logic is necessary to avoid an additional call to `judy_key` for efficiency.